pkg/reconciler/delivery/policy.go

```diff
@@ -0,0 +1,55 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package delivery
+
+import (
+	"sort"
+	"fmt"
+)
+
+
+// Policy represents the rollout strategy used to update Route objects
+type Policy struct {
+	// Mode specifies the metric that the policy is based on
+	// Possible values are: "time", "request", "error"
+	Mode string
+
+	// Percents specifies the traffic percentages that the NEW Revision is expected to have
+	// at successive rollout stages; the list of integers must start at 0
+	// all entries must be in the range [0, 100), and must be sorted in increasing order
+	// Technically the final rollout percentage is 100, but this is implicitly understood,
+	// and should NOT be explicitly specified in Percents
+	Percents []int
+
+	// Threshold specifies the conditions under which the rollout progresses to the next stage
+	// Interpretation of the Threshold value depends on the value of the Mode field
+	// For time based rollout, Threshold is the number of seconds between successive Route updates
+	// For request based rollout, Threshold is the number of requests completed
+	// For error based rollout/rollback, Threshold is the number of errors recorded
+	Threshold int
+}
+
+// ComputeNewPercent calculates, given a Policy and the current rollout stage,
+// the traffic percentage for the NEW Revision in the next rollout stage
+func ComputeNewPercent(p *Policy, oldPercent int) (int, error) {
```

Comment:
Nit: what do you think about `currentPercent` instead of `oldPercent`?

Reply:
Sounds better than oldPercent because in my comments I said "current rollout stage" :)